src/number.rs

```diff
@@ -0,0 +1,518 @@
+use std::{ ops, fmt, f32, f64 };
+use std::num::FpCategory;
+use util::grisu2;
+use util::write::write;
+
+/// NaN value represented in `Number` type. NaN is equal to itself.
+pub const NAN: Number = Number {
+    category: NAN_MASK,
+    mantissa: 0,
+    exponent: 0
+};
+
+const NEGATIVE: u8 = 0;
+const POSITIVE: u8 = 1;
+const NAN_MASK: u8 = !1;
+
+/// Number representation used inside `JsonValue`. You can easily convert
+/// the `Number` type into native Rust number types and back, or use the
+/// equality operator with another number type.
+///
+/// ```
+/// # use json::number::Number;
+/// let foo: Number = 3.14.into();
+/// let bar: f64 = foo.into();
+///
+/// assert_eq!(foo, 3.14);
+/// assert_eq!(bar, 3.14);
+/// ```
+///
+/// More often than not you will deal with `JsonValue::Number` variant that
+/// wraps around this type, instead of using the methods here directly.
+#[derive(Copy, Clone, Debug)]
+pub struct Number {
+    // A byte describing the sign and NaN-ness of the number.
+    // category == 0 -> negative sign
+    // category == 1 -> positive sign
+    // category >  1 -> NaN
+    category: u8,
```

Comment:
Might be worth mentioning the constants declared for this value at the top of the module.